core/drivers/tzc400.c

```diff
@@ -0,0 +1,420 @@
+	assert(sec_attr <= TZC_REGION_S_RDWR);
+
+	/*
+	 * Inputs look ok, start programming registers.
+	 * All the address registers are 32 bits wide and have a LOW and HIGH
+	 * component used to construct a up to a 64bit address.
+	 */
+	tzc_write_region_base_low(tzc.base, region,
+				  addr_low(region_base));
+	tzc_write_region_base_high(tzc.base, region,
+				   addr_high(region_base));
+
+	tzc_write_region_top_low(tzc.base, region,
+				addr_low(region_top));
+	tzc_write_region_top_high(tzc.base, region,
+				addr_high(region_top));
+
+	/* Assign the region to a filter and set secure attributes */
+	tzc_write_region_attributes(tzc.base, region,
+		(sec_attr << REG_ATTR_SEC_SHIFT) | filters);
+
+	/*
+	 * Specify which non-secure devices have permission to access this
+	 * region.
+	 */
+	tzc_write_region_id_access(tzc.base, region, ns_device_access);
+}
+
+
+void tzc_set_action(tzc_action_t action)
+{
+	assert(tzc.base);
+
+	/*
+	 * - Currently no handler is provided to trap an error via interrupt
+	 *   or exception.
+	 * - The interrupt action has not been tested.
+	 */
+	tzc_write_action(tzc.base, action);
+}
+
+
+void tzc_enable_filters(void)
+{
+	uint32_t state;
+	uint32_t filter;
+
+	assert(tzc.base);
+
+	for (filter = 0; filter < tzc.num_filters; filter++) {
+		state = tzc_get_gate_keeper(tzc.base, filter);
+		if (state) {
+			/* The TZC filter is already configured. Changing the
+			 * programmer's view in an active system can cause
+			 * unpredictable behavior therefore panic for now rather
+			 * than try to determine whether this is safe in this
+			 * instance. See:
+			 * http://infocenter.arm.com/help/index.jsp?\
+			 * topic=/com.arm.doc.ddi0504c/CJHHECBF.html */
+			EMSG("TZC : Filter %d Gatekeeper already enabled.\n",
+				filter);
+			panic();
+		}
+		tzc_set_gate_keeper(tzc.base, filter, 1);
+	}
+}
+
+
+void tzc_disable_filters(void)
+{
+	uint32_t filter;
+
+	assert(tzc.base);
+
+	/*
+	 * We don't do the same state check as above as the Gatekeepers are
+	 * disabled after reset.
+	 */
+	for (filter = 0; filter < tzc.num_filters; filter++)
+		tzc_set_gate_keeper(tzc.base, filter, 0);
+}
+
+#if TRACE_LEVEL >= TRACE_DEBUG
+
+static uint32_t tzc_read_region_attributes(vaddr_t base, uint32_t region)
+{
+	return read32(base + REGION_ATTRIBUTES_OFF + REGION_NUM_OFF(region));
+}
+
+static uint32_t tzc_read_region_base_low(vaddr_t base, uint32_t region)
+{
+	return read32(base + REGION_BASE_LOW_OFF + REGION_NUM_OFF(region));
+}
+
+static uint32_t tzc_read_region_base_high(vaddr_t base, uint32_t region)
+{
+	return read32(base + REGION_BASE_HIGH_OFF + REGION_NUM_OFF(region));
+}
+
+static uint32_t tzc_read_region_top_low(vaddr_t base, uint32_t region)
+{
+	return read32(base + REGION_TOP_LOW_OFF + REGION_NUM_OFF(region));
+}
+
+static uint32_t tzc_read_region_top_high(vaddr_t base, uint32_t region)
+{
+	return read32(base + REGION_TOP_HIGH_OFF + REGION_NUM_OFF(region));
+}
+
+#define	REGION_MAX		8
+static const char* tzc_attr_msg[] = {
```

Comment:
This will be caught by the Travis build, so I think you should add a line to build with all the `CFG_*` that this PR is adding (note: Travis CI is down currently but I'm investigating -- I have submitted a support request).